### PR TITLE
Add FIDO2 Attestation and MDS Validation Functionality

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Proposed changes in this pull request
This PR includes implementations on FIDO metadata service related functionality. This will allow capability to validate FIDO2 device metadata during the security key registration.
Below functionalities are introduced with this PR.
- FIDO2 attestation validations at device registration.
- FIDO2 mds validations at device registration.
- Related metadata service class and scheduler service to initialise it.
- Related configurations to enable/ disable mds service.
- Utilisation of config management API to store tenant wise configurations (can disable attestation and mds validations tenant wise)

### When should this PR be merged

Please do NOT merge until all the dependent implementations are completed and merged.

### Related issues
- https://github.com/wso2/product-is/issues/13495

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java [hunk 1]
```diff
@@ -325,6 +363,63 @@ public void finishFIDO2Registration(String challengeResponse) throws FIDO2Authen
             throw new FIDO2AuthenticatorClientException(message, FIDO2AuthenticatorConstants.ClientExceptionErrorCodes.
                     ERROR_CODE_FINISH_REGISTRATION_INVALID_REQUEST.getErrorCode());
         } else {
+            // Perform webauthn4j attestation validations if enabled.
+            if (getAuthenticatorConfigs().isAttestationValidationEnabled()) {
+                Set<String> transports = response.getCredential().getResponse().getTransports().stream().map(
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java:368]: Check whether there's a possibility for null pointer exception. In that case handle it
Check other places as well

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+ * Helper class for FIDO2 metadata validations.
+ */
+public class MetadataService {
+
+    private static final Log log = LogFactory.getLog(MetadataService.class);
+    private ObjectConverter objectConverter = null;
+    private DefaultCertPathTrustworthinessValidator defaultCertPathTrustworthinessValidator = null;
+    private static ArrayList<String> mdsEndpoints = null;
+
+    public MetadataService() {
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:60]: Remove empty constructor

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            Path[] metadataPaths = Files.list(Paths.get(readMetadataStatementDirectory())).toArray(Path[]::new);
+
+            LocalFilesMetadataStatementsProvider localFilesMetadataStatementsProvider =
+                    new LocalFilesMetadataStatementsProvider(objectConverter, metadataPaths);
+            metadataStatementsBasedTrustAnchorRepository = new MetadataStatementsBasedTrustAnchorRepository(
+                    localFilesMetadataStatementsProvider
+            );
+        } catch (IOException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Exception in constructing file based MDS blob provider: " + e.getMessage());
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:103]: Use `log.error`

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            }
+        }).filter(Objects::nonNull).toArray(MetadataBLOBProvider[]::new);
+
+        MetadataBLOBBasedTrustAnchorRepository metadataBLOBBasedTrustAnchorRepository =
+                new MetadataBLOBBasedTrustAnchorRepository(fidoMDS3MetdataBLOBProviders);
+
+        // Create local file based MDS provider (Requires to provide metadata from json files).
+        MetadataStatementsBasedTrustAnchorRepository metadataStatementsBasedTrustAnchorRepository = null;
+        try {
+            Path[] metadataPaths = Files.list(Paths.get(readMetadataStatementDirectory())).toArray(Path[]::new);
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:94]: Check whether files exists

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+     * @return DefaultCertPathTrustworthinessValidator
+     */
+    public DefaultCertPathTrustworthinessValidator getDefaultCertPathTrustworthinessValidator() {
+
+        return defaultCertPathTrustworthinessValidator;
+    }
+
+    private X509Certificate getMDS3RootCertificate() throws CertificateException, FileNotFoundException {
+
+        CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:137]: Check whether constant is defined for `X509`

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    public MetadataService() {
+    }
+
+    /**
+     * Initialize the DefaultCertPathTrustworthinessValidator object needed for webauthn4j mds validations.
+     */
+    public void initializeDefaultCertPathTrustworthinessValidator() {
+
+        objectConverter = new ObjectConverter();
+
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:69]: call `getMDS3RootCertificate()` somewhere here

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    private String readMDSRootCertificatePath() {
+
+        String value = IdentityUtil.getProperty(FIDO2AuthenticatorConstants.FIDO_MDS_ROOT_CERTIFICATE);
+
+        if (StringUtils.isNotBlank(value)) {
+            return value;
+        } else {
+            return "";
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:166]: Use String utils empty variable

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java [hunk 8]
```diff
@@ -901,4 +1029,102 @@ private String getTenantQualifiedUsername() {
...
+    }
+
+    private FIDO2Configuration getAuthenticatorConfigs() {
+
+        boolean attestationValidationEnabled;
+        boolean mdsValidationEnabled;
+
+        try {
+            attestationValidationEnabled = Boolean.parseBoolean(FIDO2AuthenticatorServiceDataHolder.getInstance()
+                    .getConfigurationManager().getAttribute(FIDO2AuthenticatorConstants.FIDO_CONFIG_RESOURCE_TYPE_NAME,
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java:1121]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java [hunk 9]
```diff
@@ -901,4 +1029,102 @@ private String getTenantQualifiedUsername() {
...
+                            FIDO2AuthenticatorConstants.FIDO2_CONFIG_ATTESTATION_VALIDATION_ATTRIBUTE_NAME
+                    ).getValue()
+            );
+        } catch (ConfigurationManagementException e) {
+            if (log.isDebugEnabled() && !Objects.equals(e.getErrorCode(),
+                    ConfigurationConstants.ErrorMessages.ERROR_CODE_ATTRIBUTE_DOES_NOT_EXISTS.getCode())) {
+                log.debug("Error in retrieving FIDO2 attestation validation configuration. " +
+                        "Using default config. Error: " + e.getMessage());
+            }
+            attestationValidationEnabled = FIDO2AuthenticatorConstants
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java:1132]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java [hunk 10]
```diff
@@ -901,4 +1029,102 @@ private String getTenantQualifiedUsername() {
...
+                            FIDO2AuthenticatorConstants.FIDO2_CONFIG_MDS_VALIDATION_ATTRIBUTE_NAME
+                    ).getValue()
+            );
+        } catch (ConfigurationManagementException e) {
+            if (log.isDebugEnabled() && !Objects.equals(e.getErrorCode(),
+                    ConfigurationConstants.ErrorMessages.ERROR_CODE_ATTRIBUTE_DOES_NOT_EXISTS.getCode())) {
+                log.debug("Error in retrieving FIDO2 mds validation configuration. " +
+                        "Using default config. Error: " + e.getMessage());
+            }
+            mdsValidationEnabled = FIDO2AuthenticatorConstants.FIDO2_CONFIG_MDS_VALIDATION_DEFAULT_VALUE;
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java:1149]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnService.java:368]: There's no need to add NPE checks as the yubico variables are annotated with `@NonNull`.
For example refer: https://github.com/Yubico/java-webauthn-server/blob/master/webauthn-server-core/src/main/java/com/yubico/webauthn/data/AuthenticatorAttestationResponse.java#L69-L73

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 2 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:60]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 3 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:103]: Fixed

Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:103]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 4 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:94]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 6 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:69]: Fixed.
Will throw an exception if the root certificate format is invalid or the file is not found.

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 7 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:166]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    private static final class MetadataInitializationTask implements Runnable {
+
+        @SneakyThrows(FIDO2AuthenticatorServerException.class)
+        @Override
+        public void run() {
+
+            log.debug("Start running the FIDO2 Metadata Initialization task.");
+            if (FIDOUtil.isMetadataValidationsEnabled()) {
+                if (FIDO2AuthenticatorServiceDataHolder.getInstance().getMetadataService() == null) {
+                    log.debug("Setting a new MetadataService object as the FIDO Authenticator metadata service " +
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java:63]: add isDebugEnabled() if check

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                        .initializeDefaultCertPathTrustworthinessValidator();
+
+                if (FIDO2AuthenticatorServiceDataHolder.getInstance().getMetadataService()
+                        .getDefaultCertPathTrustworthinessValidator() == null) {
+                    log.error("Error initializing default cert path trustworthiness validator");
+                } else {
+                    log.info("FIDO2 Metadata Initialization is successful.");
+                }
+            }
+            log.debug("Stop running the FIDO2 Metadata Initialization task.");
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java:78]: add isDebugEnabled() if check

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                            "is null");
+                    FIDO2AuthenticatorServiceDataHolder.getInstance().setMetadataService(new MetadataService());
+                }
+
+                FIDO2AuthenticatorServiceDataHolder.getInstance().getMetadataService()
+                        .initializeDefaultCertPathTrustworthinessValidator();
+
+                if (FIDO2AuthenticatorServiceDataHolder.getInstance().getMetadataService()
+                        .getDefaultCertPathTrustworthinessValidator() == null) {
+                    log.error("Error initializing default cert path trustworthiness validator");
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java:73]: Add "." at end

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java (hunk 5 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataService.java:163]: Not defined for the required case

File: pom.xml [hunk 14]
```diff
@@ -272,12 +270,32 @@
...
+                <artifactId>jackson-datatype-jsr310</artifactId>
+                <version>${jackson.datatype.jsr310.version}</version>
+            </dependency>
         </dependencies>
     </dependencyManagement>
 
     <properties>
         <carbon.kernel.package.import.version.range>[4.4.0, 5.0.0)</carbon.kernel.package.import.version.range>
-        <identity.framework.version>5.20.66</identity.framework.version>
+        <identity.framework.version>5.20.341-SNAPSHOT</identity.framework.version>
```
Comment (ThaminduDilshan) [inline pom.xml:298]:
```suggestion
        <identity.framework.version>5.20.300</identity.framework.version>
```

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java (hunk 11 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java (hunk 12 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java (hunk 13 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.fido2/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/internal/MetadataSchedulerService.java]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.fido2/pom.xml [hunk 15]
```diff
@@ -75,6 +75,18 @@
             <groupId>org.wso2.orbit.yubico.webauthn</groupId>
             <artifactId>yubico-webauthn</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.org.webauthn4j</groupId>
```
Comment (ThaminduDilshan) [inline components/org.wso2.carbon.identity.application.authenticator.fido2/pom.xml:79]:
```suggestion
            <groupId>org.wso2.orbit.webauthn4j</groupId>
```

File: features/org.wso2.carbon.identity.application.authenticator.fido2.server.feature/pom.xml [hunk 16]
```diff
@@ -74,6 +74,16 @@
             <version>${project.version}</version>
             <type>war</type>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.org.webauthn4j</groupId>
```
Comment (ThaminduDilshan) [inline features/org.wso2.carbon.identity.application.authenticator.fido2.server.feature/pom.xml:78]:
```suggestion
            <groupId>org.wso2.orbit.webauthn4j</groupId>
```

File: features/org.wso2.carbon.identity.application.authenticator.fido2.server.feature/pom.xml [hunk 17]
```diff
@@ -155,6 +165,8 @@
                                 <bundleDef>com.fasterxml.jackson.core:jackson-databind</bundleDef>
                                 <bundleDef>com.fasterxml.jackson.core:jackson-annotations</bundleDef>
                                 <bundleDef>com.fasterxml.jackson.core:jackson-core</bundleDef>
+                                <bundleDef>org.wso2.orbit.org.webauthn4j:webauthn4j</bundleDef>
```
Comment (ThaminduDilshan) [inline features/org.wso2.carbon.identity.application.authenticator.fido2.server.feature/pom.xml:168]:
```suggestion
                                <bundleDef>org.wso2.orbit.webauthn4j:webauthn4j</bundleDef>
```

File: pom.xml [hunk 18]
```diff
@@ -272,12 +270,32 @@
                 <artifactId>cxf-rt-rs-service-description</artifactId>
                 <version>${org.apache.cxf.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.wso2.orbit.org.webauthn4j</groupId>
```
Comment (ThaminduDilshan) [inline pom.xml:274]:
```suggestion
                <groupId>org.wso2.orbit.webauthn4j</groupId>
```

**Review by ashensw — APPROVED** [review]

**Pull request merged**